### PR TITLE
[Php80] Add implements interface support on single file on AddParamBasedOnParentClassMethodRector

### PR DESCRIPTION
**Pull request by samsonasik**
Based on @MartinMystikJonas comment on https://github.com/rectorphp/rector/issues/7296#issuecomment-1184287622, which `AddParamBasedOnParentClassMethodRector` doesn't add param on implements interface yet.

### CODE REVIEW

**Comment by samsonasik** [comment]
implemented 🎉 , it currently detect interface on same file, for interface in different file, that's probably require different PR.

**Comment by samsonasik** [comment]
129 PHPStan errors seems unrelated:

```bash

 ------------------------------------------------------------------------
  packages/BetterPhpDocParser/Comment/CommentsMerger.php:40
 ------------------------------------------------------------------------
  - '#Parameters should use "PhpParser\\Node\\Stmt\\ClassMethod\|PhpParser\\Node\\Expr\\Array_" types as the only types passed to this method#'
  
[ERROR] Found 129 errors      
```

ref https://github.com/rectorphp/rector-src/runs/7345146986?check_suite_focus=true 

Registered to `ignoreErrors` config https://github.com/rectorphp/rector-src/pull/2660/commits/637c07bee3ca7a0d558481ea66e9a4c43c3f7963

File: src/PhpParser/AstResolver.php [hunk 1]
```diff
@@ -105,14 +106,15 @@ public function resolveClassMethodFromMethodReflection(MethodReflection $methodR
             return null;
         }
 
-        $class = $this->betterNodeFinder->findFirstInstanceOf($nodes, Class_::class);
-        if (! $class instanceof Class_) {
+        $classLike = $this->betterNodeFinder->findFirstInstanceOf($nodes, ClassLike::class);
```
Comment (samsonasik) [inline src/PhpParser/AstResolver.php]: `findFirstInstanceOf` seems not really a solution if interface and class position are flipped:

```php
class ImplementsInterface implements InterfaceWithParam {
    public function execute()
    {
    }
}

/**
 * Both interface and class on same file on purpose to ensure locate
 * parent interface bug
 */
interface InterfaceWithParam
{
    public function execute($foo);
}
```

Comment (samsonasik) [reply src/PhpParser/AstResolver.php]: I added another failing test case for flipped posision https://github.com/rectorphp/rector-src/pull/2660/commits/932e3af361a1b48eeb3dc2d1ee0a935b42b7472b

**Comment by MartinMystikJonas** [comment]
How it would work with multiple interfaces and/or parent class and interface?  For params it should add parameter type that accepts union of types from all interfaces and parent class.

For return type rule I would need to create intersection of types from all interfaces and parent class.

**Comment by samsonasik** [comment]
that's should be working if interface resolver solved, as currently will work on multiple classes https://getrector.org/demo/511766ad-5f7f-429a-898f-5d84e2b06417

flipped position still not working on multiple classes in same file as well https://getrector.org/demo/260dfe2e-eeeb-44ce-930c-da1b5f7f5690

**Comment by samsonasik** [comment]
@MartinMystikJonas feel free to create different PR if you found a solution for it

**Comment by samsonasik** [comment]
Fixed 🎉 with loop found `ClassLike` instances to get the same `ClassLike` for multiple `ClassLike` in single file. https://github.com/rectorphp/rector-src/pull/2660/commits/0819d0b3aaf2331564834c9ba520246b1cc5466b

**Comment by samsonasik** [comment]
it seems cause error in `ReturnTypeDeclarationRector`:

```bash
There were 2 failures:

1) Rector\Tests\TypeDeclaration\Rector\FunctionLike\ReturnTypeDeclarationRector\ReturnTypeDeclarationRectorTest::test with data set #25 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/TypeDeclaration/Rector/FunctionLike/ReturnTypeDeclarationRector/Fixture/skip_respect_children_return_type_mixed.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 
 class SkipRespectChildrenReturnTypeParentMixed
 {
-    public function run()
+    public function run(): void
```

I will check more.

**Comment by samsonasik** [comment]
Fixed 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by samsonasik** [comment]
Finally 🎉 , implements from interface with different file is also working 🎉 🎉 🎉, added fixture for it  https://github.com/rectorphp/rector-src/pull/2660/commits/f824fab41db34f478e1a14b4edf8182959d73abe

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by MartinMystikJonas** [comment]
Nice. Now when we have all method definitions from parent and interfaces I think I can merge it into union type that would satisfy all. After this is merged I can try to improve it with this merging. I will also use similar mechanism for detection of intersect return type in my new rule that adds return types based on parents. I already have basic draft but it does not work yet

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Pull request merged**